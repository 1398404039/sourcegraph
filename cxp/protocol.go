package cxp

import (
	"encoding/json"

	"github.com/sourcegraph/go-langserver/pkg/lsp"
	"github.com/sourcegraph/jsonx"
	"github.com/sourcegraph/sourcegraph/xlang/lspext"
)

// ClientProxyInitializeParams contains the parameters for the client's "initialize" request to the
// CXP proxy.
//
// It is lspext.ClientProxyInitializeParams with an added nested initializationOptions.settings
// field.
type ClientProxyInitializeParams struct {
	InitializationOptions ClientProxyInitializationOptions `json:"initializationOptions"`
	Capabilities          ClientCapabilities               `json:"capabilities"`
	lspext.ClientProxyInitializeParams
}

// ClientProxyInitializeParams contains the initialization options for the client's "initialize"
// request to the CXP proxy.
type ClientProxyInitializationOptions struct {
	lspext.ClientProxyInitializationOptions
	Settings ExtensionSettings `json:"settings"`
}

// InitializeParams contains the parameters for the client's (or proxy's) "initialize" request to
// the extension.
//
// It is lspext.InitializeParams with an added nested initializationOptions.settings field.
type InitializeParams struct {
	InitializationOptions *InitializationOptions `json:"initializationOptions,omitempty"`
	Capabilities          ClientCapabilities     `json:"capabilities"`
	lspext.InitializeParams
}

// InitializationOptions contains arbitrary initialization options at the top level, plus extension
// settings.
type InitializationOptions struct {
	Other    map[string]interface{} `json:"-"`
	Settings ExtensionSettings      `json:"settings"`
}

// UnmarshalJSON implements json.Unmarshaler.
func (o *InitializationOptions) UnmarshalJSON(data []byte) error {
	var s struct {
		Settings ExtensionSettings `json:"settings"`
	}
	if err := json.Unmarshal(data, &s); err != nil {
		return err
	}
	*o = InitializationOptions{Settings: s.Settings}

	var m map[string]interface{}
	if err := json.Unmarshal(data, &m); err != nil {
		return err
	}
	delete(m, "settings")

	if len(m) > 0 {
		(*o).Other = make(map[string]interface{}, len(m))
	}
	for k, v := range m {
		(*o).Other[k] = v
	}
	return nil
}

// MarshalJSON implements json.Marshaler.
func (o InitializationOptions) MarshalJSON() ([]byte, error) {
	m := make(map[string]interface{}, len(o.Other)+1)
	for k, v := range o.Other {
		m[k] = v
	}
	m["settings"] = o.Settings
	return json.Marshal(m)
}

// ExtensionSettings contains the global/organization/user settings for an extension.
type ExtensionSettings struct {
	Merged *json.RawMessage `json:"merged,omitempty"`
}

type ClientCapabilities struct {
	lsp.ClientCapabilities

	Decorations *DecorationsCapabilityOptions `json:"decorations,omitempty"`

	// TODO(sqs): add this to cxp-js
	Exec bool `json:"exec"`
}

type InitializeResult struct {
	Capabilities ServerCapabilities `json:"capabilities"`
	lsp.InitializeResult
}

type ServerCapabilities struct {
	lsp.ServerCapabilities

	DecorationsProvider *DecorationsProviderServerCapabilities `json:"decorationsProvider,omitempty"`

	Contributions *Contributions `json:"contributions,omitempty"`
}

type DecorationsCapabilityOptions struct {
	Static  bool `json:"static,omitempty"`
	Dynamic bool `json:"dynamic,omitempty"`
}

type DecorationsProviderServerCapabilities struct {
	DecorationsCapabilityOptions
}

type TextDocumentPublishDecorationsParams struct {
	TextDocument lsp.TextDocumentIdentifier      `json:"textDocument"`
	Decorations  []lspext.TextDocumentDecoration `json:"decorations"`
}

// ParseClientCapabilities parses the client capabilities from the client's initialize message.
func ParseClientCapabilities(initializeParams []byte) (*ClientCapabilities, error) {
	var params struct {
		Capabilities ClientCapabilities `json:"capabilities"`
	}
	if err := json.Unmarshal(initializeParams, &params); err != nil {
		return nil, err
	}
	return &params.Capabilities, nil
}

type ConfigurationUpdateParams struct {
	Path  jsonx.Path  `json:"path"`
	Value interface{} `json:"value"`
}
