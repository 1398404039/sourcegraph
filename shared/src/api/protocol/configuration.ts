/**
 * A key path that refers to a location in a JSON document.
 *
 * Each successive array element specifies an index in an object or array to descend into. For example, in the
 * object `{"a": ["x", "y"]}`, the key path `["a", 1]` refers to the value `"y"`.
 */
export type KeyPath = (string | number)[]

export interface ConfigurationUpdateParams {
    /** The key path to the value. */
    path: KeyPath

    /** The new value to insert at the key path. */
    value: any
}
