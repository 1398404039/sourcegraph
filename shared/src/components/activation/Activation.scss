.activation-status-dropdown-button {
    &--hidden {
        display: none;
    }

    &__progress-bar-container {
        display: inline-block;
        width: 1rem;
        margin-left: 0.4rem;

        .CircularProgressbar-trail {
            stroke: $dropdown-border-color;
        }
        .CircularProgressbar-path {
            stroke: $oc-green-4;
        }
    }
}

.activation-status-dropdown-header {
    padding: 0.25rem 0.5rem;
}

.activation-list {
    margin-top: 0.5rem;
}

.activation-list__item {
    padding-left: 0.25rem;
    padding-right: 0.25rem;
    padding-top: 0.5rem;
    padding-bottom: 0.5rem;

    border-bottom: 1px solid $dropdown-border-color;

    @include hover-focus {
        color: $dropdown-link-hover-color;
        text-decoration: none;
        @include gradient-bg($dropdown-link-hover-bg);
    }
    &:last-child {
        border-bottom: none;
    }
}

.activation-item {
    cursor: pointer;
    padding-left: 0.25rem;
    padding-right: 0.25rem;
    padding-top: 0.25rem;
    padding-bottom: 0.25rem;

    &__checkbox {
        &--done {
            color: $oc-green-6;
        }
        &--todo {
            color: $dropdown-border-color;
        }
    }
}

.first-use-button {
    display: block;
    position: relative;
}

.first-use-button::before,
.first-use-button::after {
    position: absolute;
    content: '';
    display: block;
    width: 140%;
    height: 100%;
    left: -20%;
    z-index: 1000;
    transition: all ease-in-out 0.5s;
    background-repeat: no-repeat;
}
.first-use-button::before {
    display: none;
    top: -75%;
    background-image: radial-gradient(circle, #00b4f2 20%, transparent 20%),
        radial-gradient(circle, transparent 12%, #f96216 20%, transparent 30%),
        radial-gradient(circle, #f96216 20%, transparent 20%), radial-gradient(circle, #00b4f2 20%, transparent 20%),
        radial-gradient(circle, transparent 10%, #b200f8 15%, transparent 20%),
        radial-gradient(circle, #b200f8 20%, transparent 20%), radial-gradient(circle, #b200f8 20%, transparent 20%),
        radial-gradient(circle, #00b4f2 20%, transparent 20%), radial-gradient(circle, #f96216 20%, transparent 20%);
    background-size: 10% 20%, 20% 40%, 15% 40%, 20% 50%, 18% 40%, 10% 10%, 15% 30%, 10% 50%, 18% 18%;
}
.first-use-button::after {
    display: none;
    bottom: -75%;
    background-image: radial-gradient(circle, #00b4f2 20%, transparent 20%),
        radial-gradient(circle, #f96216 20%, transparent 20%),
        radial-gradient(circle, transparent 8%, #00b4f2 15%, transparent 20%),
        radial-gradient(circle, #b200f8 20%, transparent 20%), radial-gradient(circle, #00b4f2 20%, transparent 20%),
        radial-gradient(circle, #00b4f2 20%, transparent 20%), radial-gradient(circle, #b200f8 20%, transparent 20%);
    background-size: 15% 30%, 20% 50%, 18% 40%, 20% 50%, 15% 30%, 10% 20%, 20% 40%;
}
.first-use-button.animate::before {
    display: block;
    animation: topBubbles ease-in-out 0.75s forwards;
}
.first-use-button.animate::after {
    display: block;
    animation: bottomBubbles ease-in-out 0.75s forwards;
}

@keyframes topBubbles {
    0% {
        background-position: 5% 120%, 10% 120%, 10% 120%, 15% 120%, 25% 120%, 25% 100%, 40% 100%, 55% 100%, 70% 100%;
    }
    50% {
        background-position: 0% 100%, 0% 50%, 10% 80%, 20% 0%, 30% 30%, 22% 50%, 50% 50%, 65% 20%, 90% 30%;
    }
    100% {
        background-position: 0% 70%, 0% 100%, 10% 100%, 20% -50%, 30% 70%, 22% 40%, 50% 80%, 65% 10%, 90% 70%;
        background-size: 0% 0%, 0% 0%, 0% 0%, 0% 0%, 0% 0%, 0% 0%;
    }
}
@keyframes bottomBubbles {
    0% {
        background-position: 10% -40%, 30% 10%, 55% -10%, 70% -10%, 85% -10%, 70% -10%, 70% 0%;
    }
    50% {
        background-position: 0% 80%, 20% 80%, 5% 60%, 60% 100%, 75% 70%, 95% 60%, 105% 0%;
    }
    100% {
        background-position: -40% 140%, 20% 140%, 45% 140%, 60% 140%, 75% 130%, 95% 70%, 120% 100%;
        background-size: 0% 0%, 0% 0%, 0% 0%, 0% 0%, 0% 0%, 0% 0%;
    }
}
