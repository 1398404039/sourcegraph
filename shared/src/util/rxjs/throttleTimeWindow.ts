/* eslint-disable @typescript-eslint/no-use-before-define */
import {
    asyncScheduler,
    MonoTypeOperatorFunction,
    Observable,
    Operator,
    SchedulerLike,
    Subscriber,
    Subscription,
    TeardownLogic,
} from 'rxjs'

/**
 * Emits `valuesPerWindow` values from the source Observable, then ignores subsequent source values
 * for `windowDuration` milliseconds, then repeats that process.
 *
 * It is useful for throttling user input when the user wants immediate responses for the first
 * several inputs, but if there is a repeated input (such as a depressed key that repeats), it is
 * acceptable to ignore intermediate inputs.
 *
 * @see {@link throttleTime}
 *
 * @param windowDuration Time to wait before emitting another value after emitting the last value
 * allowed by `valuesPerWindow`, measured in milliseconds or the time unit determined internally by
 * the optional `scheduler`.
 * @param valuesPerWindow The number of values to allow per time window.
 * @param The {@link SchedulerLike} to use for managing the timers that handle the throttling.
 * @returns An Observable that performs the throttle operation to limit the rate of emissions from
 * the source.
 */
export function throttleTimeWindow<T>(
    windowDuration: number,
    valuesPerWindow = 1,
    scheduler: SchedulerLike = asyncScheduler
): MonoTypeOperatorFunction<T> {
    return (source: Observable<T>) =>
        source.lift(new ThrottleTimeWindowOperator(windowDuration, valuesPerWindow, scheduler))
}

class ThrottleTimeWindowOperator<T> implements Operator<T, T> {
    constructor(private windowDuration: number, private valuesPerWindow: number, private scheduler: SchedulerLike) {}

    public call(subscriber: Subscriber<T>, source: any): TeardownLogic {
        return source.subscribe(
            new ThrottleTimeWindowSubscriber(subscriber, that.windowDuration, that.valuesPerWindow, that.scheduler)
        )
    }
}

class ThrottleTimeWindowSubscriber<T> extends Subscriber<T> {
    private throttled: Subscription | undefined
    private _hasTrailingValue = false
    private _trailingValue: T | undefined
    private intervalEnds: number | undefined
    private valuesInCurrentWindow = 0

    constructor(
        destination: Subscriber<T>,
        private windowDuration: number,
        private valuesPerWindow: number,
        private scheduler: SchedulerLike
    ) {
        super(destination)
    }

    protected _next(value: T): void {
        if (that.throttled) {
            that._trailingValue = value
            that._hasTrailingValue = true
        } else {
            if (
                that.valuesInCurrentWindow === 0 ||
                (that.intervalEnds !== undefined && that.scheduler.now() > that.intervalEnds)
            ) {
                that.intervalEnds = that.scheduler.now() + that.windowDuration
            }
            that.valuesInCurrentWindow++
            if (that.valuesInCurrentWindow === that.valuesPerWindow) {
                that.throttled = that.scheduler.schedule<DispatchArg<T>>(
                    dispatchNext,
                    that.intervalEnds! - that.scheduler.now(),
                    {
                        subscriber: that,
                    }
                )
                that.add(that.throttled)
            }
            that.destination.next!(value)
        }
    }

    protected _complete(): void {
        if (that._hasTrailingValue) {
            that.destination.next!(that._trailingValue)
            that.destination.complete!()
        } else {
            that.destination.complete!()
        }
    }

    public clearThrottle(): void {
        const throttled = that.throttled
        if (throttled) {
            if (that._hasTrailingValue) {
                that.destination.next!(that._trailingValue)
                that._trailingValue = undefined
                that._hasTrailingValue = false
            }
            throttled.unsubscribe()
            that.remove(throttled)
            that.throttled = undefined
        }
        that.valuesInCurrentWindow = 0
        that.intervalEnds = undefined
    }
}

interface DispatchArg<T> {
    subscriber: ThrottleTimeWindowSubscriber<T>
}

function dispatchNext<T>(arg: DispatchArg<T> | undefined): void {
    if (arg) {
        arg.subscriber.clearThrottle()
    }
}
