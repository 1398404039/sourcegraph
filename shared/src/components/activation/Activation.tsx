import H from 'history'
import { BehaviorSubject, Observable } from 'rxjs'
import { first } from 'rxjs/operators'

export interface ActivationStep {
    id: string
    title: string
    detail: string
    action: (h: H.History) => void
}

/**
 * Tracks the activation status of the current user.
 */
export class ActivationStatus {
    private steps_: ActivationStep[]
    private fetchCompleted: () => Observable<{ [key: string]: boolean }>
    private completed_: BehaviorSubject<{ [key: string]: boolean } | null>

    /**
     * Accepts the activation steps to track and a function that fetches the completed
     * activation steps (in the form of a function that returns an Observable).
     */
    constructor(steps: ActivationStep[], fetchCompleted: () => Observable<{ [key: string]: boolean }>) {
        this.steps_ = steps
        this.fetchCompleted = fetchCompleted
        this.completed_ = new BehaviorSubject<{ [key: string]: boolean } | null>(null)
    }

    public get steps(): ActivationStep[] {
        return this.steps_
    }

    /**
     * The completion status. Null indicates that the completion status has never been
     * fetched from the server.
     */
    public get completed(): BehaviorSubject<{ [key: string]: boolean } | null> {
        return this.completed_
    }

    public ensureInit(): void {
        if (!this.completed.value) {
            this.update(null)
        }
    }

    /**
     * Updates the current completion status with the partial completition status `u`.
     * If `u` is null, refetches the completion status using `fetchCompleted`.
     * If the current completion status is null and `u` is non-null, has no effect.
     * Must be called with null parameter first (this triggers the first fetch of
     * the completion status via `fetchCompleted`).
     */
    public update(u: { [key: string]: boolean } | null): void {
        if (!u) {
            this.fetchCompleted()
                .pipe(first()) // subscription will get auto-cleaned up
                .subscribe(res => {
                    const nextCompleted: { [key: string]: boolean } = {}
                    for (const step of this.steps) {
                        nextCompleted[step.id] = res[step.id] || false
                    }
                    this.completed.next(nextCompleted)
                })
        } else {
            if (!this.completed.value) {
                return
            }
            const newVal: { [key: string]: boolean } = {}
            Object.assign(newVal, this.completed.value)
            for (const step of this.steps) {
                if (u[step.id] !== undefined) {
                    newVal[step.id] = u[step.id]
                }
            }
            this.completed.next(newVal)
        }
    }
}

/**
 * Returns the percent of activation checklist items completed.
 */
export const percentageDone = (info?: { [key: string]: boolean }): number => {
    if (!info) {
        return 0
    }
    const vals = Object.values(info)
    return (100 * vals.filter(e => e).length) / vals.length
}
