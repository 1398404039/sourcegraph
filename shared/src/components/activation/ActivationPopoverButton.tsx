import H from 'history'
import RocketIcon from 'mdi-react/RocketIcon'
import React from 'react'
import CircularProgressbar from 'react-circular-progressbar'
import ButtonDropdown from 'reactstrap/lib/ButtonDropdown'
import DropdownItem from 'reactstrap/lib/DropdownItem'
import DropdownMenu from 'reactstrap/lib/DropdownMenu'
import DropdownToggle from 'reactstrap/lib/DropdownToggle'
import { Subscription } from 'rxjs'
import { Props as CommandListProps } from '../../commandPalette/CommandList'
import { PopoverButton } from '../PopoverButton'
import { ActivationStatus, percentageDone } from './Activation'
import { ActivationChecklist, ActivationChecklistItem, ActivationChecklistProps } from './ActivationChecklist'

export interface ActivationDropdownProps {
    history: H.History
    activation: ActivationStatus
}

interface State {
    isOpen: boolean

    animate: boolean
    show?: boolean
    completed?: { [key: string]: boolean }
}

export class ActivationDropdown extends React.PureComponent<ActivationDropdownProps, State> {
    public state: State = { isOpen: false, animate: false }
    private subscriptions = new Subscription()

    private toggleIsOpen = () => this.setState(prevState => ({ isOpen: !prevState.isOpen }))

    public componentWillMount(): void {
        this.subscriptions.add(
            this.props.activation.completed.subscribe(completed => {
                if (!completed) {
                    // Completion is unloaded
                    this.setState({ show: false })
                    return
                }
                if (!this.state.completed && completed && percentageDone(completed) >= 100) {
                    // Completion was previously unloaded and now is 100% done
                    this.setState({ show: false })
                    return
                }
                if (
                    completed &&
                    this.state.completed &&
                    percentageDone(completed) > percentageDone(this.state.completed)
                ) {
                    // Completition was loaded and is being updated
                    this.setState({ completed, animate: true })
                    setTimeout(() => {
                        if (percentageDone(this.state.completed) >= 100) {
                            this.setState({ animate: false, show: false })
                        } else {
                            this.setState({ animate: false, show: true })
                        }
                    }, 1500)
                    return
                }

                // Completion is being updated (catch all)
                this.setState({ completed, show: true })
            })
        )
        this.props.activation.ensureInit()
    }
    public componentWillUnmount(): void {
        this.subscriptions.unsubscribe()
    }
    public render(): JSX.Element {
        return (
            <ButtonDropdown
                isOpen={this.state.isOpen}
                toggle={this.toggleIsOpen}
                className={`${this.state.show ? '' : 'activation-status-dropdown-button--hidden'} nav-link p-0`}
            >
                <DropdownToggle
                    caret={false}
                    className={`${
                        this.state.animate ? 'animate' : ''
                    } first-use-button bg-transparent d-flex align-items-center e2e-user-nav-item-toggle`}
                    nav={true}
                >
                    Setup
                    <span className="activation-status-dropdown-button__progress-bar-container">
                        <CircularProgressbar strokeWidth={12} percentage={percentageDone(this.state.completed)} />
                    </span>
                </DropdownToggle>
                <DropdownMenu right={true}>
                    <DropdownItem header={true} className="py-1">
                        <div className="activation-status-dropdown-header">
                            <h2>
                                <RocketIcon className="icon-inline" /> Hi there!
                            </h2>
                            <div>Complete the steps below to finish onboarding to Sourcegraph.</div>
                        </div>
                    </DropdownItem>
                    <DropdownItem divider={true} />
                    {this.state.completed ? (
                        this.props.activation.steps.map(s => (
                            <div key={s.id} className="dropdown-item" onClick={this.toggleIsOpen}>
                                <ActivationChecklistItem
                                    {...s}
                                    history={this.props.history}
                                    done={(this.state.completed && this.state.completed[s.id]) || false}
                                />
                            </div>
                        ))
                    ) : (
                        <div>Loading...</div>
                    )}
                </DropdownMenu>
            </ButtonDropdown>
        )
    }
}

// interface ActivationPopoverDropdownProps extends ActivationChecklistProps {
//     onClick: () => void
// }

// /**
//  * Presents the site admin activation checklist in a navbar dropdown.
//  */
// export class ActivationPopoverDropdown extends React.PureComponent<ActivationPopoverDropdownProps> {
//     public render(): JSX.Element {
//         return (
//             <div
//                 className="activation-container command-list list-group list-group-flush rounded"
//                 onClick={this.props.onClick}
//             >
//                 <div className="list-group-item">
//                     <div className="activation-container__header">
//                         <h2>
//                             <RocketIcon className="icon-inline" /> Hi there!
//                         </h2>
//                         <div>Complete the steps below to finish onboarding to Sourcegraph.</div>
//                     </div>
//                 </div>
//                 <div className="list-group-item">{<ActivationChecklist {...this.props} />}</div>
//             </div>
//         )
//     }
// }

// export interface ActivationPopoverButtonProps extends CommandListProps {
//     history: H.History
//     activation: ActivationStatus
// }
// export interface ActivationPopoverButtonState {
//     animate: boolean
//     show?: boolean
//     completed?: { [key: string]: boolean }

//     dropdownHideOnChange?: any // controls dismissing the dropdown
// }

// /**
//  * The nav bar button that displays the current user activation status (percentage of
//  * activation steps completed and a dropdown list of all activation items and their
//  * completion status).
//  */
// export class ActivationPopoverButton extends React.PureComponent<
//     ActivationPopoverButtonProps,
//     ActivationPopoverButtonState
// > {
//     private subscriptions = new Subscription()
//     constructor(props: ActivationPopoverButtonProps) {
//         super(props)
//         this.state = { animate: false }
//     }
//     public componentWillMount(): void {
//         this.subscriptions.add(
//             this.props.activation.completed.subscribe(completed => {
//                 if (!completed) {
//                     // Completion is unloaded
//                     this.setState({ show: false })
//                     return
//                 }
//                 if (!this.state.completed && completed && percentageDone(completed) >= 100) {
//                     // Completion was previously unloaded and now is 100% done
//                     this.setState({ show: false })
//                     return
//                 }
//                 if (
//                     completed &&
//                     this.state.completed &&
//                     percentageDone(completed) > percentageDone(this.state.completed)
//                 ) {
//                     // Completition was loaded and is being updated
//                     this.setState({ completed, animate: true })
//                     setTimeout(() => {
//                         if (percentageDone(this.state.completed) >= 100) {
//                             this.setState({ animate: false, show: false })
//                         } else {
//                             this.setState({ animate: false, show: true })
//                         }
//                     }, 2000)
//                     return
//                 }

//                 // Completion is being updated (catch all)
//                 this.setState({ completed, show: true })
//             })
//         )
//         this.props.activation.ensureInit()
//     }
//     public componentWillUnmount(): void {
//         this.subscriptions.unsubscribe()
//     }
//     private onClick = (e: React.MouseEvent<HTMLElement, MouseEvent>) => {
//         e.preventDefault()
//     }
//     private onClickDropdown = () => {
//         this.setState({ dropdownHideOnChange: !this.state.dropdownHideOnChange })
//     }
//     public render(): JSX.Element | null {
//         return (
//             <div className={`activation-status-dropdown-button-container ${this.state.show ? '' : 'hidden'}`}>
//                 <PopoverButton
//                     className={`first-use-button activation-status-dropdown-button ${
//                         this.state.animate ? 'animate' : ''
//                     }`}
//                     {...this.state}
//                     popoverClassName="rounded"
//                     placement="auto-end"
//                     hideOnChange={this.state.dropdownHideOnChange}
//                     hideCaret={true}
//                     popoverElement={
//                         <ActivationPopoverDropdown
//                             steps={this.props.activation.steps}
//                             history={this.props.history}
//                             onClick={this.onClickDropdown}
//                             {...this.state}
//                         />
//                     }
//                 >
//                     <a className="nav-link activation-status-dropdown-button__nav-link" href="#" onClick={this.onClick}>
//                         Setup
//                         <span className="activation-status-dropdown-button__progress-bar-container">
//                             <CircularProgressbar strokeWidth={12} percentage={percentageDone(this.state.completed)} />
//                         </span>
//                     </a>
//                 </PopoverButton>
//             </div>
//         )
//     }
// }
