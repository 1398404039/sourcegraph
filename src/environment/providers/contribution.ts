import { BehaviorSubject, Observable, Unsubscribable } from 'rxjs'
import { distinctUntilChanged, map } from 'rxjs/operators'
import { ContributableMenu, Contributions, MenuItemContribution } from '../../protocol'
import { isEqual } from '../../util'

/** A registered set of contributions from an extension in the registry. */
export interface ContributionsEntry {
    /** The contributions. */
    contributions: Contributions
}

/** Manages and executes contributions from all extensions. */
export class ContributionRegistry {
    private _entries = new BehaviorSubject<ContributionsEntry[]>([])

    public registerContributions(entry: ContributionsEntry): Unsubscribable {
        this._entries.next([...this._entries.value, entry])
        return {
            unsubscribe: () => {
                this._entries.next(this._entries.value.filter(e => e !== entry))
            },
        }
    }

    /**
     * All contributions (merged), emitted whenever the set of registered contributions changes.
     */
    public readonly contributions: Observable<Contributions> = this.getContributions(this._entries)

    protected getContributions(entries: Observable<ContributionsEntry[]>): Observable<Contributions> {
        return entries.pipe(
            map(entries => mergeContributions(entries.map(e => e.contributions))),
            distinctUntilChanged((a, b) => isEqual(a, b))
        )
    }

    /**
     * The current set of contributions (merged). Used by callers that do not need to react to contributions being
     * registered or unregistered.
     */
    public get contributionsSnapshot(): Contributions {
        return mergeContributions(this._entries.value.map(e => e.contributions))
    }

    /**
     * All contribution entries, emitted whenever the set of registered contributions changes.
     *
     * Most callers should use ContributionsRegistry#contributions. Only use #entries if the caller needs
     * information that is discarded when the contributions are merged (such as the extension that registered each
     * set of contributions).
     */
    public readonly entries: Observable<ContributionsEntry[]> & { value: ContributionsEntry[] } = this._entries
}

/**
 * Merges the contributions.
 *
 * Most callers should use ContributionRegistry's contributions field, which merges all registered contributions.
 */
export function mergeContributions(contributions: Contributions[]): Contributions {
    if (contributions.length === 0) {
        return {}
    }
    if (contributions.length === 1) {
        return contributions[0]
    }
    const merged: Contributions = {}
    for (const c of contributions) {
        if (c.commands) {
            if (!merged.commands) {
                merged.commands = [...c.commands]
            } else {
                merged.commands = [...merged.commands, ...c.commands]
            }
        }
        if (c.menus) {
            if (!merged.menus) {
                merged.menus = { ...c.menus }
            } else {
                for (const [menu, items] of Object.entries(c.menus) as [ContributableMenu, MenuItemContribution[]][]) {
                    if (!merged.menus[menu]) {
                        merged.menus[menu] = [...items]
                    } else {
                        merged.menus[menu] = [...merged.menus[menu]!, ...items]
                    }
                }
            }
        }
    }
    return merged
}
