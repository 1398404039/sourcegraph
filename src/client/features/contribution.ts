import { Subscription } from 'rxjs'
import uuidv4 from 'uuid/v4'
import { ContributionRegistry } from '../../environment/providers/contribution'
import { MessageType as RPCMessageType } from '../../jsonrpc2/messages'
import { ClientCapabilities, Contributions, ServerCapabilities } from '../../protocol'
import { DynamicFeature, ensure, RegistrationData } from './common'

/**
 * Support for user-facing features contributed by the server for use and display in the client's application.
 */
export class ContributionFeature implements DynamicFeature<Contributions> {
    private contributions = new Map<string, Subscription>()

    constructor(private registry: ContributionRegistry) {}

    public get messages(): RPCMessageType {
        // This method is not actually a protocol method, but some value is required here because the client tracks
        // DynamicFeatures based on their method name.
        return { method: 'window/contribution' }
    }

    public fillClientCapabilities(capabilities: ClientCapabilities): void {
        ensure(ensure(capabilities, 'window')!, 'contribution')!.dynamicRegistration = true
    }

    public initialize(capabilities: ServerCapabilities): void {
        if (!capabilities.contributions) {
            return
        }
        this.register(this.messages, {
            id: uuidv4(),
            registerOptions: capabilities.contributions,
        })
    }

    public register(_message: RPCMessageType, data: RegistrationData<Contributions>): void {
        const existing = this.contributions.has(data.id)
        if (existing && !data.overwriteExisting) {
            throw new Error(`registration already exists with ID ${data.id}`)
        }
        if (data.overwriteExisting) {
            if (!existing) {
                throw new Error(`no existing registration to overwrite with ID ${data.id}`)
            }
            this.unregister(data.id)
        }
        const sub = new Subscription()
        sub.add(this.registry.registerContributions({ contributions: data.registerOptions }))
        this.contributions.set(data.id, sub)
    }

    public unregister(id: string): void {
        const sub = this.contributions.get(id)
        if (!sub) {
            throw new Error(`no registration with ID ${id}`)
        }
        sub.unsubscribe()
        this.contributions.delete(id)
    }

    public unregisterAll(): void {
        for (const sub of this.contributions.values()) {
            sub.unsubscribe()
        }
        this.contributions.clear()
    }
}
