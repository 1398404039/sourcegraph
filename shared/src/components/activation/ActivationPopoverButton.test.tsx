import * as H from 'history'
import React from 'react'
import renderer from 'react-test-renderer'
import { of } from 'rxjs'
import { ContributableMenu } from '../../api/protocol/contribution'
import { ActivationStatus } from './Activation'
import { ActivationPopoverButton } from './ActivationPopoverButton'

describe('ActivationPopoverButton', () => {
    const NOOP_PLATFORM_CONTEXT = { forceUpdateTooltip: () => void 0 }
    const NOOP_EXTENSIONS_CONTROLLER = { services: {} as any, executeCommand: async () => void 0 }

    test('render 0/2', () => {
        const activation = new ActivationStatus(
            [
                {
                    id: 'id1',
                    title: 'title1',
                    detail: 'detail1',
                    action: (h: H.History) => void 0,
                },
                {
                    id: 'id2',
                    title: 'title2',
                    detail: 'detail2',
                    action: (h: H.History) => void 0,
                },
            ],
            () => of({})
        )
        const history = H.createMemoryHistory({ keyLength: 0 })
        const component = renderer.create(
            <ActivationPopoverButton
                history={history}
                activation={activation}
                location={history.location}
                menu={ContributableMenu.GlobalNav}
                extensionsController={NOOP_EXTENSIONS_CONTROLLER}
                platformContext={NOOP_PLATFORM_CONTEXT}
            />
        )
        expect(component.toJSON()).toMatchSnapshot()
    })
    test('render 1/2', () => {
        const activation = new ActivationStatus(
            [
                {
                    id: 'id1',
                    title: 'title1',
                    detail: 'detail1',
                    action: (h: H.History) => void 0,
                },
                {
                    id: 'id2',
                    title: 'title2',
                    detail: 'detail2',
                    action: (h: H.History) => void 0,
                },
            ],
            () => of({ id1: true })
        )
        const history = H.createMemoryHistory({ keyLength: 0 })
        const component = renderer.create(
            <ActivationPopoverButton
                history={history}
                activation={activation}
                location={history.location}
                menu={ContributableMenu.GlobalNav}
                extensionsController={NOOP_EXTENSIONS_CONTROLLER}
                platformContext={NOOP_PLATFORM_CONTEXT}
            />
        )
        expect(component.toJSON()).toMatchSnapshot()
    })
})
