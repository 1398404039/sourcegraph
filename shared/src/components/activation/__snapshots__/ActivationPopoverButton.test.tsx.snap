// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivationPopoverButton render 0/2 1`] = `
<div>
  <div
    className="popover-button2  activation-status-dropdown-button popover-button2__btn popover-button2__anchor"
  >
    <span
      className="popover-button2__container"
      onClick={[Function]}
    >
      <span
        className="link-like"
      >
        Setup
      </span>
      <div
        className="activation-status-dropdown-button__progress-bar-container "
      >
        <svg
          className="CircularProgressbar "
          viewBox="0 0 100 100"
        >
          <path
            className="CircularProgressbar-trail"
            d="
      M 50,50
      m 0,-44
      a 44,44 0 1 1 0,88
      a 44,44 0 1 1 0,-88
    "
            fillOpacity={0}
            strokeWidth={12}
            style={
              Object {
                "stroke": "#ff8787",
              }
            }
          />
          <path
            className="CircularProgressbar-path"
            d="
      M 50,50
      m 0,-44
      a 44,44 0 1 1 0,88
      a 44,44 0 1 1 0,-88
    "
            fillOpacity={0}
            strokeWidth={12}
            style={
              Object {
                "stroke": "#69db7c",
                "strokeDasharray": "276.46015351590177px 276.46015351590177px",
                "strokeDashoffset": "276.46015351590177px",
              }
            }
          />
        </svg>
      </div>
       
    </span>
  </div>
</div>
`;

exports[`ActivationPopoverButton render 1/2 1`] = `
<div>
  <div
    className="popover-button2  activation-status-dropdown-button popover-button2__btn popover-button2__anchor"
  >
    <span
      className="popover-button2__container"
      onClick={[Function]}
    >
      <span
        className="link-like"
      >
        Setup
      </span>
      <div
        className="activation-status-dropdown-button__progress-bar-container "
      >
        <svg
          className="CircularProgressbar "
          viewBox="0 0 100 100"
        >
          <path
            className="CircularProgressbar-trail"
            d="
      M 50,50
      m 0,-44
      a 44,44 0 1 1 0,88
      a 44,44 0 1 1 0,-88
    "
            fillOpacity={0}
            strokeWidth={12}
            style={
              Object {
                "stroke": "#ff8787",
              }
            }
          />
          <path
            className="CircularProgressbar-path"
            d="
      M 50,50
      m 0,-44
      a 44,44 0 1 1 0,88
      a 44,44 0 1 1 0,-88
    "
            fillOpacity={0}
            strokeWidth={12}
            style={
              Object {
                "stroke": "#69db7c",
                "strokeDasharray": "276.46015351590177px 276.46015351590177px",
                "strokeDashoffset": "138.23007675795088px",
              }
            }
          />
        </svg>
      </div>
       
    </span>
  </div>
</div>
`;
